modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/sorted/PageMemorySortedIndexStorage.java

```diff
@@ -0,0 +1,138 @@
+    private final IndexColumnsFreeList freeList;
+
+    /** Sorted index tree instance. */
+    private final SortedIndexTree sortedIndexTree;
+
+    /** Partition id. */
+    private final int partitionId;
+
+    /**
+     * Constructor.
+     *
+     * @param descriptor Sorted index descriptor.
+     * @param freeList Free list to store index columns.
+     * @param sortedIndexTree Sorted index tree instance.
+     */
+    public PageMemorySortedIndexStorage(SortedIndexDescriptor descriptor, IndexColumnsFreeList freeList, SortedIndexTree sortedIndexTree) {
+        this.descriptor = descriptor;
+        this.freeList = freeList;
+        this.sortedIndexTree = sortedIndexTree;
+
+        partitionId = sortedIndexTree.partitionId();
+    }
+
+    @Override
+    public SortedIndexDescriptor indexDescriptor() {
+        return descriptor;
+    }
+
+    @Override
+    public void put(IndexRow row) {
+        IndexColumns indexColumns = new IndexColumns(partitionId, row.indexColumns().byteBuffer());
+
+        try {
+            SortedIndexRow sortedIndexRow = new SortedIndexRow(indexColumns, row.rowId());
+
+            var insert = new InsertSortedIndexRowInvokeClosure(sortedIndexRow, freeList, IoStatisticsHolderNoOp.INSTANCE);
+
+            sortedIndexTree.invoke(sortedIndexRow, null, insert);
+        } catch (IgniteInternalCheckedException e) {
+            throw new StorageException("Failed to put value into index", e);
+        }
+    }
+
+    @Override
+    public void remove(IndexRow row) {
+        IndexColumns indexColumns = new IndexColumns(partitionId, row.indexColumns().byteBuffer());
+
+        try {
+            SortedIndexRow hashIndexRow = new SortedIndexRow(indexColumns, row.rowId());
+
+            var remove = new RemoveSortedIndexRowInvokeClosure(hashIndexRow, freeList, IoStatisticsHolderNoOp.INSTANCE);
+
+            sortedIndexTree.invoke(hashIndexRow, null, remove);
+
+            // Performs actual deletion from freeList if necessary.
+            remove.afterCompletion();
+        } catch (IgniteInternalCheckedException e) {
+            throw new StorageException("Failed to remove value from index", e);
+        }
+    }
+
+    @Override
+    public Cursor<IndexRow> scan(@Nullable BinaryTuplePrefix lowerBound, @Nullable BinaryTuplePrefix upperBound, int flags) {
+        IgniteCursor<SortedIndexRow> cursor;
+
+        try {
+            cursor = sortedIndexTree.find(
+                    toSortedIndexRowKey(lowerBound),
+                    toSortedIndexRowKey(upperBound),
+                    (flags & GREATER_OR_EQUAL) != 0,
+                    (flags & LESS_OR_EQUAL) != 0,
+                    null,
+                    null
+            );
+        } catch (IgniteInternalCheckedException e) {
+            throw new StorageException("Failed to create scan cursor", e);
+        }
+
+        return Cursor.fromIterator(new TreeCursorAdapter<>(cursor, sortedIndexRow -> new IndexRow() {
```

Comment:
There exists a `IndexRowImpl` class, no need for the anonymous class here

Reply:
Fix it